File: typescript/packages/bench/src/concerns/hasMetadata.ts
/** Ordered key/value metadata. */

export interface HasMetadata {
    metadata(): ReadonlyMap<string, string>;
    meta(key: string): string | undefined;
    metaKeys(): readonly string[];
}

/** Insertion-ordered bag, so metadata rendering is deterministic. */
export class MetadataBag implements HasMetadata {
    private readonly entries = new Map<string, string>();

    set(key: string, value: string): this {
        this.entries.set(key, value);

        return this;
    }

    metadata(): ReadonlyMap<string, string> {
        return this.entries;
    }

    meta(key: string): string | undefined {
        return this.entries.get(key);
    }

    metaKeys(): readonly string[] {
        return [...this.entries.keys()];
    }
}

File: rust/crates/atelier-bench/src/concerns/has_metadata.rs
//! Ordered key/value metadata.

use indexmap::IndexMap;

pub trait HasMetadata {
    fn metadata(&self) -> &IndexMap<String, String>;

    fn meta(&self, key: &str) -> Option<&str> {
        self.metadata().get(key).map(String::as_str)
    }

    fn meta_keys(&self) -> Vec<&str> {
        self.metadata().keys().map(String::as_str).collect()
    }
}

/// Insertion-ordered bag, so metadata rendering is deterministic.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct MetadataBag {
    entries: IndexMap<String, String>,
}

impl MetadataBag {
    pub fn set(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
        self.entries.insert(key.into(), value.into());
        self
    }
}

impl HasMetadata for MetadataBag {
    fn metadata(&self) -> &IndexMap<String, String> {
        &self.entries
    }
}
